### ISSUE TITLE
checkpoint for writer

### ISSUE THREAD

**Issue by mincong-h** [mincong-h/gsoc-hsearch#110]

**Comment by mincong-h** [comment]
`org.hibernate.search.backend.AddLuceneWork` is not serializable, so I can't give it to checkpoint for the item writer.

**Comment by mincong-h** [comment]
According to the _JSR-352-1.0-Final-Release.pdf_, §11.13 Stop Processing : 

> ## 11.13 Stop Processing

> The JobOperator.stop operation stops a running job execution. If a step is running at the time the stop is invoked, the batch runtime takes the following actions:

> ### Chunk Step

> The job and step batch status is marked STOPPING. Note the batch runtime cannot guarantee the step actually exits. The batch runtime attempts to interrupt the read/process/write chunk processing loop. The batch runtime allows the step to finish processing the current item. This means the current item is read, processed if a processor is configured, and all currently buffered items, if any, including the current item, are written. If the batch artifacts configured on the chunk type step return to the batch runtime, as expected, the job and step batch status is marked STOPPED.

> ### Batchlet Step

> ...

From the chunk step paragraph, we can see that the last item read is ensured to be written before the job stop. In the same time, this item's ID will be saved as the reader-checkpoint. So, the job can safely restarted from that reader-checkpoint. Therefore, there's no need to add a checkpoint for the writer.

@emmanuelbernard has mentioned the extreme situations, e.g shutdown of the JVM, interruption of power supply. In those cases, job instance will be end with batch status _FAILED_  instead of _STOPPED_. And for these situations, we need to redo the job from the very beginning. This is described in _JSR-352-1.0-Fianl-Release_, page 46 : 

Value | Meaning
--- | ---
STOPPED | Batch job has been stopped through the JobOperator interface stop operation or by a <stop> element in the Job XML. A step has a status of STOPPED once it has actually been stopped by the batch runtime.
FAILED | Batch job has ended due to an unresolved exception or by a <fail> element in the Job XML. A step has a status of FAILED under the same conditions.

In conclusion, I think checkpoint in item-reader is good enough and there's no need to do another checkpoint in item-writer.
